src/Ocelot/Configuration/UpstreamRoutingHeaders.cs

```diff
@@ -0,0 +1,70 @@
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Ocelot.Configuration
+{
+    public class UpstreamRoutingHeaders
+    {
+        public IReadOnlyDictionary<string, HashSet<string>> Headers { get; }
+
+        public UpstreamRoutingHeaders(IReadOnlyDictionary<string, HashSet<string>> headers)
+        {
+            Headers = headers;
+        }
+
+        public bool Any() => Headers.Any();
+
+        public bool HasAnyOf(IHeaderDictionary requestHeaders)
+        {
+            IHeaderDictionary lowerCaseHeaders = GetLowerCaseHeaders(requestHeaders);
+            foreach (KeyValuePair<string, HashSet<string>> h in Headers)
+            {
+                if (lowerCaseHeaders.TryGetValue(h.Key, out var values))
+                {
+                    HashSet<string> requestHeaderValues = new(values);
+                    if (h.Value.Overlaps(requestHeaderValues))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public bool HasAllOf(IHeaderDictionary requestHeaders)
+        {
+            IHeaderDictionary lowerCaseHeaders = GetLowerCaseHeaders(requestHeaders);
+            foreach (KeyValuePair<string, HashSet<string>> h in Headers)
+            {
+                if (!lowerCaseHeaders.TryGetValue(h.Key, out var values))
+                {
+                    return false;
+                }
+
+                HashSet<string> requestHeaderValues = new(values);
+                if (!h.Value.Overlaps(requestHeaderValues))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static IHeaderDictionary GetLowerCaseHeaders(IHeaderDictionary headers)
+        {
+            IHeaderDictionary lowerCaseHeaders = new HeaderDictionary();
+            foreach (KeyValuePair<string, StringValues> kv in headers)
+            {
+                string key = kv.Key.ToLowerInvariant();
+                StringValues values = new(kv.Value.Select(v => v.ToLowerInvariant()).ToArray());
```

Comment:
Because header values from the configuration are also converted to lowercase in the creator and the implementation uses `HashSet` to do an overlap test without case-insensitive comparer.

Reply:
Actually, two points on this:
* According to [docs](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings), `ToUpper` is recommended for string normalization
* I will use case-insensitive `StringComparer` for header value comparison so a conversion like this won't be necessary (I'll still have to convert the header name because I need to do the `TryGetValue` on the header dictionary).

Reply:
Yeah, this stupid situation with `TryGetValue` for key comparisons is really ugly.
Well... It is your design to convert to upper case.
Thanks for fixing it in https://github.com/ThreeMammals/Ocelot/pull/1684/commits/6d29040db6570a01847ef596a2cdc583e146ad6c!